### ISSUE TITLE
How to visualize attention weights?

### ISSUE THREAD

**Issue by hadaev8** [idiap/fast-transformers#20]
Would be very useful, thx.

**Comment by angeloskath** [comment]
Hi,

Yep, this is in the TODO list. Like everything else, the planned solution aims to be general. I am leaning towards an event system that would allow for zero overhead and completely decoupled capturing of intermediate tensors.

I 'll ping you again when the feature branch is created so you can monitor its progress.

Cheers,
Angelos

**Comment by maxbaluev** [comment]
@angeloskath Very necessary functionality, how far is it in roadmap?

**Comment by angeloskath** [comment]
Hi Max,

It is not a very hard feature but there are things I want to streamline first. I am working on having continuous integration and perhaps even continuous deployment so that I don't have to go through the error prone manual build > test > sdist > pypi_upload procedure.

Not more than a week. The feature is going to be the following:

1. There is going to be an optional event dispatcher passed to all modules which when set to None it will be using a global event dispatcher
2. We will define some events and all modules will be dispatching said events together with the id of the layer (same as for the state_dict)

When one wants to plot the attention of the n-th layer then we can just listen to that event and be done with it. If for instance you want to add an entropy regularizer to the attention it should something as simple as the following pseudocode.

```python
import torch
from fast_transformer.events import EventDispatcher, AttentionEvent

class AttentionRegularizer:
    def __init__(self):
        EventDispatcher.default.listen(AttentionEvent, self.regularize_attention)
        self.entropy = 0

    def regularize_attention(self, event):
        self.entropy += -(torch.log(event.attention_matrix + 1e-6) * event.attention_matrix).sum(-1).mean()
```

Comments are welcome! Thanks for your patience!

Cheers,
Angelos

**Comment by hadaev8** [comment]
@angeloskath any updates?

**Comment by angeloskath** [comment]
Hi,

check out the branch (that I literally just pushed) [event-dispatcher](https://github.com/idiap/fast-transformers/tree/event-dispatcher). It is far from usable. It doesn't have a single test and the event-dispatcher is not integrated with the transformer/attention modules. Other than that it is a simple little observer pattern implementation.

I know that it is not moving fast enough and I am sorry for that.

Cheers,
Angelos

**Comment by angeloskath** [comment]
@hadaev8 It should be fully usable now. I haven't merged into master yet but you could pip install in dev mode and check it out if you want. You can see the `tests/events/test_events.py` for some example usage or the snippet below.

```python
from fast_transformers.builders import TransformerEncoderBuilder
# Import the event dispatcher in order to attach a listener and the attention event
# to specify what to listen for
from fast_transformers.events import EventDispatcher, AttentionEvent
import torch

# Make a small event listener class to accumulate the attention
# matrices in a list
class AttentionCollector:
    def __init__(self):
        self.matrices = []

    # Any callable is an event listener
    def __call__(self, event):
        self.matrices.append(event.attention_matrix)

    def clear(self):
        self.matrices = []

# Attach it to the dispatcher
attentions = AttentionCollector()
EventDispatcher.get().listen(AttentionEvent, attentions)

# Ok let's make a transformer and run some dummy inputs
transformer = TransformerEncoderBuilder().get()
transformer(torch.rand(10, 100, 4*64))
print([a.shape for a in attentions.matrices])
attentions.clear()
print([a.shape for a in attentions.matrices])
transformer(torch.rand(10, 100, 4*64))
print([a.shape for a in attentions.matrices])
```

**Comment by angeloskath** [comment]
Ok, so I wrote the docs and merged it into master (the docs are not published yet since there is no new release yet). The list of events is not complete but currently softmax attention does emit the attention matrix and the attention layer does emit the queries, keys and values.

This means I am closing this issue, if anybody has questions/problems with the event system, feel free to reopen it.

Cheers,
Angelos

**Comment by hadaev8** [comment]
@angeloskath 
Should you advice how to install it?
Guess, i can clone repo and run setup, but i dont know what command to run.

```
usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
   or: setup.py --help [cmd1 cmd2 ...]
   or: setup.py --help-commands
   or: setup.py cmd --help

error: no commands supplied
```

**Comment by angeloskath** [comment]
Sure, you can do the following,

```
cd fast-transformers/
pip install -e .  # do a local install in dev mode
```

and if you change something in C++ and want to rebuild you can run

```
python setup.py build_ext --inplace
```

I should probably add a "Develop/Contribute" section in the docs.

Cheers,
Angelos

**Comment by angeloskath** [comment]
Oh and if you want to see the docs for whatever is in the repo right now you can run

```
pip install pdoc
pip install mkdocs
```
to install the dependencies for building the documentation website and then
```
./tools.py serve_docs  # and then point your browser to localhost:8000
```
to build and serve the website locally.

**Comment by hadaev8** [comment]
@angeloskath 
Thanks for your efforts, would you support recurrent and linear attention?

**Comment by angeloskath** [comment]
Sure! It's just a bit more complicated since for instance for linear attention the attention matrix is never explicitly computed.

However, I am indeed planning a series of events that will be lazy, as in the attention matrix will be computed on the fly if requested from the queries, keys and values. This will allow for accessing the attention matrix of Linear or Clustered attentions.

For now, for linear you could compute the attention from the QKVEvent. It is little more than the following (untested code):

```python
qk = torch.einsum("nlhe,nshe->nhls", event.queries, event.keys)
z = torch.einsum("nlhe,nhe->nl", event.queries, event.keys.sum(1))
attention_matrix = qk / (z[:, None, :, None] + 1e-6)
```

Cheers,
Angelos

**Comment by hadaev8** [comment]
@angeloskath 
Is it possible to aggregate all attention matrices per sample?
Now its all added to flat list.
I want to look at the attentions of random sample in the last batch on validation.

**Comment by angeloskath** [comment]
Hm, I am not sure I follow, you mean aggregate them in a large tensor? Something like the following?

```python
attentions = []
def attention_collector(event):
    attentions.append(event.attention_matrix)
attentions = torch.cat(attentions, dim=0)
```

**Comment by hadaev8** [comment]
No.
This is my toy example:
https://colab.research.google.com/drive/10zY6dX2iDm1Mz4vOk6bSzN8CYhF5Ev2l?authuser=1#scrollTo=tBXk0i8_L98V&line=3&uniqifier=1
So I need to check attentions on validation samples.
And take one random sample from last batch to plot.
I know what with 3 encoder 1 decoder i will have 5 attentions so i can generate random number and take subset of all attentions like attentions[5*random:5*(random+1)]
But it would be uncomfortable, because if i change model architecture i need to go to validation function and replace 5 with another number.
Also would be nice to track whatever attention it is encoder/decoder self/not self so i can plot it in category in tensorboard.
How should i perform it?

**Comment by angeloskath** [comment]
I see, so you can actually filter events based on the layer that emits them. Like the following code for example.

```python
from collections import defaultdict
from fast_transformers.events import EventDispatcher, AttentionEvent
from fast_transformers.events.filters import from_layer

# Get your net
net = ......

# Example that selects to listen from a specific layer
EventDispatcher.get().listen(
    from_layer(net.encoder.layers[2].attention.inner_attention),
    lambda ev: print(ev.attention_matrix)
)

# More involved example that actually creates a map of all attentions with layer names
layer_names = {m: n for n, m in net.named_modules()}
attentions = defaultdict(list)
EventDispatcher.get().listen(
    AttentionEvent,
    lambda ev: attentions[layer_names[ev.source]].append(ev.attention_matrix)
)
```

Now you can for instance get all the attention matrices from a given layer using `attentions["human.readable.name"]`.

**Comment by hadaev8** [comment]
@angeloskath 
Should I expect such attention maps?
![1](https://i.imgur.com/pB790kt.png)
I guess it's a lack of normalization.
So if I have full, linear, and casual linear, what is the best way to plot all attention maps?
